### PR TITLE
Add message dialog and directory open dialog on OS X

### PR DESCRIPTION
**Pull request by kkirby**
This implements #2471.

One issue I had was compiling both Objective-C and C++ together. Apparently Objective-C doesn't like the `-std` flag. To get around this, I split up the `c`, `cpp` and `mm` files into their own variables in the `CMakelists` file. This way the `cpp` files get the `-std` flag, and the `mm` files don't. I don't know if there is a better option. There was apparently an issue filed in CMake to implement a `OBJC_FLAGS` variable, but they didn't do it. https://cmake.org/Bug/view.php?id=4756

### CODE REVIEW

**Comment by LRFLEW** [comment]
2 things. First, there's no reason to use .mm (aka, C++ with Objective-C). It would make more sense to use .m (aka, C with Objective-C). Second, this won't be the only thing requiring Objective-C (I can tell you that right now), so it would make more sense to rename `osx.c` to `osx.m`, as it's only being compiled on OS X anyways.

File: src/platform/linux.c [hunk 1]
```diff
@@ -111,4 +111,19 @@ void platform_posix_sub_user_data_path(char *buffer, const char *homedir, const
...
+	STUB();
+	return NULL;
+}
+
+void platform_show_messagebox(char *message)
+{
+	STUB();
+	log_verbose(message);
+}
+}
```
Comment (janisozaur) [inline src/platform/linux.c]: one too many?

**Comment by LRFLEW** [comment]
Also, I'm fairly certain you messed up the linux build as you're trying to get Linux to build the Objective-C files.

**Comment by janisozaur** [comment]
@LRFLEW not necessarily, at least not for this reason. The CMake variable will just be empty, which should be fine.

**Comment by kkirby** [comment]
@LRFLEW Ah! I for some reason forgot that Objective-C is C by default, and thought `mm` was the one with Objective-C and C.

@janisozaur, good catch. I guess I forgot to compile once more after that.

**Comment by kkirby** [comment]
@LRFLEW, when renaming `osx.c` to `osx.m` and importing `Cocoa.h`, I'm getting typename redeclaration issues.

`rct2.h:50:23: error: typedef redefinition with different types ('unsigned long' vs 'uint32_t' (aka 'unsigned int'))`

`/rct2.h:46:21: error: typedef redefinition with different types ('long' vs 'int32_t' (aka 'int'))`

**Comment by LRFLEW** [comment]
I'm seeing that. It's seems to be a problem with `stdint.h`. I'm looking into it now.

**Comment by kkirby** [comment]
I've published my WIP, which doesn't compile, if you wanted something to work off of. But I have to go for now.

**Comment by LRFLEW** [comment]
I tested your version with the temporary fix and it worked just fine. I think the problem is more systemic, so this may just be the solution we need right now.

Side note: there's also `platform_open_common_file_dialog` that might want to be implemented here as well.

**Comment by kkirby** [comment]
Okay, I cleaned the code up. Can someone take a look? Thanks!

**Comment by LinusU** [comment]
~~Maybe we can fix the type declarations by defining ours the same way that cocoa is? That would eliminate the need for `NO_INT_TYPE_DECLARE`.~~ See #2480

**Comment by kkirby** [comment]
I've updated my branch and removed the conditional check. Thanks! I've also synced it with develop upstream.

**Comment by pizza2004** [comment]
I have this merged in, and don't see the dialog box pop up when trying to use the native dialog box for loading/saving, but it does when it wants you to find the RCT2 folder at first.

**Comment by kkirby** [comment]
@pizza2004 This doesn't implement the save/load dialog.

**Comment by pizza2004** [comment]
Ah, okay.  May I ask specifically where this occurs then?  Just at the beginning, when it asks where the RCT2 files are?

**Comment by kkirby** [comment]
@pizza2004, it's only used to select the rct2 install directory.

**Comment by kkirby** [comment]
This still needs a little bit of work and cleaning up. I'll let you know when it's ready to merge.

**Comment by LRFLEW** [comment]
What's the status on this? Some other things I'm working on is dependent on changing `osx.c` to `osx.m`,and I'm waiting on this for that.

**Comment by kkirby** [comment]
It's done but I need to clean up the code. I haven't had a chance lately. I can try and get to it tomorrow, or you can feel free to pull it, modify; and submit a PR to my repo. Or if you think the code is fine as-is, then I think that's probably good enough; but it's up to everyone else.

**Comment by kkirby** [comment]
@LRFLEW This should be good to go. I'd really appreciate it if you could do a quick peer review of the code. I'm a tad bit rusty on my Objective-C.

@IntelOrca, pending anyone's objections, this is good to go. Thanks!

**Comment by LRFLEW** [comment]
@kkirby Looks like there's some merge conflicts. You may need to rebase.

**Comment by LRFLEW** [comment]
@kkirby I may be wrong about this, but IIRC, the "Objective-C" way of including a framework is `@import Cocoa;`, not `#include <Cocoa/Cocoa.h>`. Also, you are including the Cocoa header, but telling the linker you need the AppKit framework. Those should probably be changed to match (preferable to whichever one doesn't include the other).

**Comment by kkirby** [comment]
@LRFLEW, thanks. I noticed that right after I sent the comment. I believe I resolved the conflicts with `CMakeList.txt` properly. I needed to add the `-framework AppKit` for OS X, but with all the changes I was unsure if I place it in the proper location.

**Comment by LRFLEW** [comment]
I don't think there's a wrong place to put the `-framework` declaration. Personally, I'd leave it where it is, as the place the other arguments were placed were specifically for adding the assembly.

**Comment by kkirby** [comment]
Good catch on the import statement. I've never seen an `@import` declaration before. I tried it out blindly and it didn't work. But I did change it to `<AppKit/AppKit.h>`.

**Comment by LRFLEW** [comment]
The `@import` statement works for me. I think it's the "correct" way to do it because it's what's used in the documentation. (see [the upper-right corner](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/ApplicationKit/Classes/NSOpenPanel_Class/index.html))

**Comment by kkirby** [comment]
Ah, figured it out. I hadn't enabled modules. I've updated the `CMakeLists.txt` file and also the `osx.m` file to use the module import. Thanks.

**Comment by janisozaur** [comment]
Please don't set `CMAKE_SHARED_LINKER_FLAGS` when you're not targeting a shared library. Use `CMAKE_EXE_LINKER_FLAGS`

**Comment by kkirby** [comment]
@janisozaur Wow. Not going to lie. I thought `SHARED` meant something totally different. Thanks for pointing that out. But why does line [59](https://github.com/OpenRCT2/OpenRCT2/blob/develop/CMakeLists.txt#L59) have `SHARED`?

**Comment by janisozaur** [comment]
Because, as I recall, it's still being used by mingw. And when I initially ported openrct2 to Linux, I was aiming for feature parity with Windows, so I made a shared object (I think you call them dylibs) and it should still be possible.

And please don't add the `-framework` where you did, use https://github.com/OpenRCT2/OpenRCT2/pull/2473/files#diff-af3b638bc2a3e6c650974192a53c7291R138 instead.

**Comment by kkirby** [comment]
I don't know if I agree with you on where the `-framework` flag should go. The location you specified is commented as:

    Handle creating the rct2 text and data files on OS X and Linux

Which the `-framework` flag is not a part of. I think maybe line [107](https://github.com/OpenRCT2/OpenRCT2/pull/2473/files#diff-af3b638bc2a3e6c650974192a53c7291L107) might be a better spot since it's adding libraries to link against. But I'm unsure. If you're insistent on me putting it where you're suggesting, then I'll respect your decision.

**Comment by janisozaur** [comment]
line

I mean this line, which for me displays as in the screenshot.

**Comment by kkirby** [comment]
Right. The entire enclosing block is all related to the functionality of creating and injecting the segments into the resulting executable.

**Comment by janisozaur** [comment]
OK, I may have not understood you well. In any case, I still mean this line, as I find our CMakeLists.txt file is already quite complicated as is and would not like to introduce more complexity when not really needed.

**Comment by kkirby** [comment]
Okay. So, If I'm understanding properly, then based on your logic, then all of [this](https://github.com/kkirby/OpenRCT2/blob/feature/dialogs/CMakeLists.txt#L59-L68) should go in there, too, correct? I'm not saying that's what you said, but I'm suggesting that's how things *should* be based on your logic. I just feel like the `-framework` flag is 100% unassociated with the segment stuff. But you're saying it should go there, so then everything that is *nix related should go in that related block as well. The only reason I'm pushing on this is because I like for things to make sense. And if someone were to go through the CMakeLists file and find the `-framework` flag in the midst of all the segment stuff it might be confusing because `AppKit` has absolutely nothing to do with the segments.

**Comment by janisozaur** [comment]
While I would like for this to happen, it unfortunately cannot, as this would break compilation, since CMake scans this file top-to-bottom. These need to stay where they are.

**Comment by janisozaur** [comment]
If you insist, please leave `-framework` where it is, it's not really that much of a deal.

But please at least do basic sanity check before you push commits, as in the very next line you overwrite what you just set.

**Comment by kkirby** [comment]
Hey, great catch! I'll do you one better. I'll remove the `-framework` flag entirely. Apparently with LRFLEW  suggestion to use modules, we don't even need the `-framework` flag. I didn't notice the overwriting of the variable because it ended up compiling and running and it hadn't before without the `-framework` flag.

**Comment by janisozaur** [comment]
As I can't read Objective-C, I haven't reviewed that. CMakeLists.txt is ok in my opinion.

**Comment by kkirby** [comment]
@LRFLEW When you get a chance, can you peer review this again? Thanks!

**Comment by LRFLEW** [comment]
There's only one thing that stands out to me, which is that you're calling `release` in one of the functions, but not in all of them. From my own experimenting in Xcode, it looks like 32-bit apps do not support ARC, which I think means you have to release everything you create (but I'm not certain about that). I'll look into it some more, but otherwise it looks fine to me.

**Comment by kkirby** [comment]
It looks like, according to this [SO Post](http://stackoverflow.com/a/8702692), it will be auto released. I actually did an analysis on the code with releases on all the objects and Xcode yelled at me to remove the releases. So, I believe it's correct the way it is.

**Comment by LRFLEW** [comment]
@kkirby If you saw my edits, you can see that I figured out that those objects are auto-released, but the question is *when* they are released. At the latest, it will be released by the time the app closes, but I'm not certain where it will exactly be released. It should probably be by the time that function exits, but I'm not certain if that's how it works by default or if we need to add an `@autoreleasepool` section to the functions.

**Comment by kkirby** [comment]
@LRFLEW, I saw your edits. The code is fine. The objects will be released at the end of an iteration of the main run loop. See [here](http://stackoverflow.com/questions/5192308/objective-c-autorelease-memory-management).

**Comment by LRFLEW** [comment]
>The objects will be released at the end of an iteration of the main run loop.

Yes, but no. Since the run loop is not going though AppKit/Cocoa, and is instead coded into C, there's no possible way for it to know *when* that is. To the Objective-C code, it thinks it's in a spin lock/infinite loop.

**Comment by kkirby** [comment]
Doesn't SDL use the main run loop? If not, I can just add an auto release block. It's not a big deal.

**Comment by LRFLEW** [comment]
Give me a minute to test it using Inspector.

**Comment by LRFLEW** [comment]
Just test it, and my concern appears to be accurate. It seems that the autorelease pool is well outside of what gets called normally. I was able to have the game actually running (so the game loop was constantly looping) and the memory was not being released. It's technically not a memory leak, but you should still use `@autoreleasepool` around all of the methods using Objective-C.

**Comment by kkirby** [comment]
Okay. I'm fine with that. Were you able to confirm that the memory was released by using an `@autoreleasepool`? I'm unsure how you tested this so I can't really test on my end.

**Comment by LRFLEW** [comment]
Yes, using an `@autoreleasepool` caused immediate deallocation.

**Comment by kkirby** [comment]
Awesome, thanks for checking on that. I really appreciate it. I've made the corresponding updates. So I know for the future, how did you test it?

**Comment by LRFLEW** [comment]
I was using the Xcode project I'm working on getting ready for merging into the project. Using Instruments' "Allocations" view, I looked for an NSAlert object while the alert was visible. I then kept track of it specifically, and checked to see when it was being removed from the list (or not in this case). In general, though, all of the functions in this file that uses any Objective-C code should probably have the `@autoreleasepool`, even if it's not apparently using any autoreleased objects.

**Comment by kkirby** [comment]
Good to know. Thanks again. Do you think this is good to merge?

**Comment by LRFLEW** [comment]
I say this looks good now. No significant complaints (maybe one or two in terms of code style, but nothing in terms of functionality).

**Comment by kkirby** [comment]
Well, lets make a standard for Objective-C. What would you change about the style? I'm 100% open to ideas. Objective-C is obviously not my native language, so I haven't adopted any kind of specific style.

**Comment by LRFLEW** [comment]
For Objective-C style, I'd suggest using the property notation when possible (so `path.UTF8String` instead of `[path UFT8String]`). Other than that, the only change I'd make is decrease your code repetition using polymorphism (so since `NSOpenPanel` is a subclass of `NSSavePanel`, you can reduce code repetition using a single `NSSavePanel` pointer). I've made a PR to your fork for what I'd change. It's functionally the same, but uses these improvements.

**Comment by kkirby** [comment]
Awesome. I like those changes. I'll wait for Travis to build and then ping IntelOrca.

**Comment by kkirby** [comment]
@IntelOrca this is good to go. Thanks!

File: src/platform/osx.m [hunk 2]
```diff
@@ -75,4 +75,77 @@ void platform_posix_sub_user_data_path(char *buffer, const char *homedir, const
...
+		if ([panel runModal] == NSFileHandlingPanelCancelButton){
+			return 0;
+		} else {
+			strcpy(filename, panel.URL.path.UTF8String);
+			return 1;
+		}
+	}
+}
+
+#endif
```
Comment (LinusU) [inline src/platform/osx.m]: [No newline at end of file](http://stackoverflow.com/a/729795)

**Comment by kkirby** [comment]
@LinusU Sorry about that. I've fixed it and re-pushed.

**Comment by LRFLEW** [comment]
@kkirby I see the change on Github just fine. This should be good to go now.

**Pull request merged**